crates/zkevm_circuits/src/bn254/ec_mul/implementation.rs

```diff
@@ -0,0 +1,428 @@
+            &negated,
+            &*y,
+        );
+    }
+
+    // now decompose every scalar we are interested in
+    let k1_msb_decomposition = to_width_4_window_form(cs, scalar_decomposition.k1);
+    let k2_msb_decomposition = to_width_4_window_form(cs, scalar_decomposition.k2);
+
+    let mut comparison_constants = Vec::with_capacity(PRECOMPUTATION_TABLE_SIZE);
+    for i in 1..=PRECOMPUTATION_TABLE_SIZE {
+        let constant = Num::allocated_constant(cs, F::from_u64_unchecked(i as u64));
+        comparison_constants.push(constant);
+    }
+
+    // now we do amortized double and add
+    let mut acc = SWProjectivePoint::zero(cs, base_field_params);
+    assert_eq!(
+        k1_msb_decomposition.len(),
+        NUM_MULTIPLICATION_STEPS_FOR_WIDTH_4
+    );
+    assert_eq!(
+        k2_msb_decomposition.len(),
+        NUM_MULTIPLICATION_STEPS_FOR_WIDTH_4
+    );
+
+    for (idx, (k1_window_idx, k2_window_idx)) in k1_msb_decomposition
+        .into_iter()
+        .zip(k2_msb_decomposition.into_iter())
+        .enumerate()
+    {
+        let ignore_k1_part = k1_window_idx.is_zero(cs);
+        let ignore_k2_part = k2_window_idx.is_zero(cs);
+
+        let (mut selected_k1_part_x, mut selected_k1_part_y) = table[0].clone();
+        let (mut selected_k2_part_x, mut selected_k2_part_y) = endomorphisms_table[0].clone();
+        for i in 1..PRECOMPUTATION_TABLE_SIZE {
+            let should_select_k1 = Num::equals(cs, &comparison_constants[i], &k1_window_idx);
+            let should_select_k2 = Num::equals(cs, &comparison_constants[i], &k2_window_idx);
+            selected_k1_part_x = Selectable::conditionally_select(
+                cs,
+                should_select_k1,
+                &table[i].0,
+                &selected_k1_part_x,
+            );
+            selected_k1_part_y = Selectable::conditionally_select(
+                cs,
+                should_select_k1,
+                &table[i].1,
+                &selected_k1_part_y,
+            );
+            selected_k2_part_x = Selectable::conditionally_select(
+                cs,
+                should_select_k2,
+                &endomorphisms_table[i].0,
+                &selected_k2_part_x,
+            );
+            selected_k2_part_y = Selectable::conditionally_select(
+                cs,
+                should_select_k2,
+                &endomorphisms_table[i].1,
+                &selected_k2_part_y,
+            );
+        }
+
+        let tmp_acc = acc.add_mixed(cs, &mut (selected_k1_part_x, selected_k1_part_y));
+        acc = Selectable::conditionally_select(cs, ignore_k1_part, &acc, &tmp_acc);
+        let tmp_acc = acc.add_mixed(cs, &mut (selected_k2_part_x, selected_k2_part_y));
+        acc = Selectable::conditionally_select(cs, ignore_k2_part, &acc, &tmp_acc);
+
+        if idx != NUM_MULTIPLICATION_STEPS_FOR_WIDTH_4 - 1 {
+            for _ in 0..WINDOW_WIDTH {
+                acc = acc.double(cs);
+            }
+        }
+    }
+
+    acc
```

Comment:
So if the point is infinity, then you mask with the generator and then return the wrong result?( You should remove this generator after). (Hint:  use condition select)